src/lib/x509/x509_builder.cpp

```diff
@@ -0,0 +1,272 @@
+
+      void add_uri(std::string_view uri) { m_uri.emplace_back(uri); }
+
+      void add_xmpp(std::string_view xmpp) { m_xmpp.emplace_back(xmpp); }
+
+      void add_ipv4(uint32_t ipv4) { m_ipv4.push_back(ipv4); }
+
+      void add_allowed_usage(Key_Constraints usage) { m_usage |= usage; }
+
+      void add_allowed_extended_usage(const OID& usage) { m_ext_usage.push_back(usage); }
+
+      void set_as_ca_certificate(std::optional<size_t> path_limit) {
+         BOTAN_STATE_CHECK(!m_is_ca_request);
+         m_is_ca_request = true;
+         m_path_limit = path_limit;
+      }
+
+   private:
+      void add_subject_dn(std::string_view attr, std::string_view value) {
+         BOTAN_ARG_CHECK(!value.empty(), "Empty DN value is invalid");
+         m_subject_dn.add_attribute(attr, value);
+      }
+
+      std::unique_ptr<Certificate_Extension> setup_alt_name(const Extensions& extensions) const {
+         AlternativeName subject_alt;
+
+         /*
+         If the extension was already created in extensions we need to merge the
+         values provided with the extension value
+         */
+         if(const auto* ext = extensions.get_extension_object_as<Cert_Extension::Subject_Alternative_Name>()) {
+            subject_alt = ext->get_alt_name();
+         }
+
+         for(const auto& dns : m_dns) {
+            subject_alt.add_dns(dns);
+         }
+         for(const auto& uri : m_uri) {
+            subject_alt.add_uri(uri);
+         }
+         for(const auto& email : m_email) {
+            subject_alt.add_email(email);
+         }
+         for(const auto& xmpp : m_xmpp) {
+            subject_alt.add_other_name(OID::from_string("PKIX.XMPPAddr"), ASN1_String(xmpp, ASN1_Type::Utf8String));
+         }
+         for(const uint32_t ipv4 : m_ipv4) {
+            subject_alt.add_ipv4_address(ipv4);
+         }
+
+         return std::make_unique<Cert_Extension::Subject_Alternative_Name>(subject_alt);
+      }
+
+      X509_DN m_subject_dn;
+      Extensions m_extensions;
+      Key_Constraints m_usage;
+      std::vector<std::string> m_email;
+      std::vector<std::string> m_dns;
+      std::vector<std::string> m_uri;
+      std::vector<std::string> m_xmpp;
+      std::vector<uint32_t> m_ipv4;
+      std::vector<OID> m_ext_usage;
+      bool m_is_ca_request = false;
+      std::optional<size_t> m_path_limit;
+};
+
+CertificateParametersBuilder::CertificateParametersBuilder() :
+      m_state(std::make_unique<CertificateParametersBuilder::State>()) {}
+
+CertificateParametersBuilder::CertificateParametersBuilder(CertificateParametersBuilder&& other) noexcept = default;
+
+CertificateParametersBuilder::~CertificateParametersBuilder() = default;
+
+X509_Certificate CertificateParametersBuilder::into_self_signed_cert(std::chrono::system_clock::time_point not_before,
+                                                                     std::chrono::system_clock::time_point not_after,
```

Comment:
This is slightly annoying, `X509_CA::sign_request` takes an `X509_Time`, this takes a time point. Would be nice to use only one

Reply:
I'll add a second version of `sign_request` taking time points. It's `X509_Time` there mostly because that API predates adoption of C++11 and `std::chrono`.... but it's generally more useful for an application to specify time as a chrono type vs our ASN.1 type

Reply:
It's fine to leave it as is, it just means one 3 line conversion function for the ffi.
I already have one anyway for `uint64_t -> X509_Time`, so adding one for `uint64_t -> time_point -> X509_Time` isn't a big deal